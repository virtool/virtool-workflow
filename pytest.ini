[pytest]
addopts = --disable-pytest-warnings --docker-compose-remove-volumes
