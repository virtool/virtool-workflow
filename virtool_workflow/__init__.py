"""An SDK for defining Virtool workflows."""
from pyfixtures import fixture
from virtool_workflow.decorators import step
from virtool_workflow.workflow import Workflow
from virtool_workflow.runtime.executor import execute
from virtool_workflow.runtime.step import WorkflowStep

__all__ = [
    "Workflow",
    "WorkflowStep",
    "execute",
    "fixture",
    "step",
]
