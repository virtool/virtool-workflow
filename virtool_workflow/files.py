from dataclasses import dataclass
from datetime import datetime
from typing import Literal

VirtoolFileFormat = Literal[
    "sam",
    "bam",
    "fasta",
    "fastq",
    "csv",
    "tsv",
    "json",
    "unknown",
]


@dataclass
class VirtoolFile:
    id: int
    name: str
    size: int
    format: VirtoolFileFormat
    name_on_disk: str = None
    uploaded_at: datetime = None
