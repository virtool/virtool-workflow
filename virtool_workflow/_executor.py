import asyncio
import logging
from asyncio import CancelledError
from contextlib import asynccontextmanager, contextmanager
from itertools import chain

from fixtures import FixtureScope, fixture
from virtool_workflow import Workflow
from virtool_workflow._steps import WorkflowStep
from virtool_workflow.events import Events
from virtool_workflow.execution import states
from virtool_workflow.hooks import (
    on_failure,
    on_finish,
    on_result,
    on_step_finish,
    on_step_start,
    on_success,
    on_workflow_start,
    on_terminated,
    on_cancelled,
    on_error,
)

logger = logging.getLogger(__name__)


async def execute(
    workflow: Workflow, scope: FixtureScope, events: Events
) -> FixtureScope:
    """
    Execute a workflow.

    :param workflow: The workflow to execute
    :param scope: The :class:`FixtureScope` to use for fixture injection
    """
    scope["step_number"] = 0
    scope["error"] = None
    scope["logger"] = logger
    scope["workflow"] = workflow
    scope["current_step"] = None

    await on_workflow_start.trigger(scope)

    try:
        with update_state_in_scope(scope):
            for step in chain(workflow.on_startup, workflow.steps, workflow.on_cleanup):
                bound_step = await scope.bind(step.function)

                async with run_step_with_hooks(scope, step):
                    logger.info(f"Running step '{step.display_name}'")
                    await bound_step()
    except CancelledError:
        if events.cancelled.is_set():
            await asyncio.gather(
                on_cancelled.trigger(scope),
                on_failure.trigger(scope),
            )
        else:
            if not events.terminated.is_set():
                logger.warning("Workflow cancelled for unknown reason")

            await asyncio.gather(
                on_terminated.trigger(scope),
                on_failure.trigger(scope),
            )
    except Exception as error:
        scope["error"] = error
        await asyncio.gather(on_error.trigger(scope), on_failure.trigger(scope))
    else:
        if "results" in scope:
            await on_result.trigger(scope)

        await on_success.trigger(scope)
    finally:
        await on_finish.trigger(scope)

    return scope


@contextmanager
def update_state_in_scope(scope: FixtureScope):
    """
    Update the scope state value as the enclosed workflow execution proceeds.

    :param scope: the workflow fixture scope
    """
    scope["state"] = states.RUNNING

    try:
        yield
    except Exception:
        scope["state"] = states.ERROR
        raise
    else:
        scope["state"] = states.COMPLETE


@asynccontextmanager
async def run_step_with_hooks(scope: FixtureScope, step: WorkflowStep):
    """
    Run the passed ``step`` while updating the ``scope`` and triggering hooks.

    :param scope: the scope to update and use for triggering hooks
    :param step: the workflow step
    """
    scope["current_step"] = step
    scope["step_number"] += 1

    try:
        await on_step_start.trigger(scope)
        yield
        await on_step_finish.trigger(scope)
    finally:
        scope["current_step"] = None


@fixture(scope="function")
def progress(step_number, workflow):
    return float(step_number) / float(len(workflow.steps))
