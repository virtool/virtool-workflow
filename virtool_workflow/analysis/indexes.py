import gzip
import json
import shutil
from dataclasses import dataclass
from pathlib import Path
from typing import Awaitable, Callable, Dict, List, Optional, Tuple

import aiofiles
from virtool_core.utils import compress_file, decompress_file
from virtool_workflow import data_model
from virtool_workflow.api.indexes import IndexProvider
from virtool_workflow.data_model.files import VirtoolFileFormat
from virtool_workflow.execution.run_in_executor import FunctionExecutor
from virtool_workflow.execution.run_subprocess import RunSubprocess

from fixtures import fixture


async def not_implemented(*args):
    raise NotImplementedError()


@dataclass
class Index(data_model.Index):
    """
    Represents a Virtool reference index.

    Use cases:


    1. Access index data when creating an analysis workflow.

       Downloads and provides access to the index JSON, Bowtie2, and FASTA at:
           - :attr:`.bowtie_path`

       Allows lookup of key index values using
           - :meth:`.get_otu_id_by_sequence_id`
           - :meth:`.get_sequence_length`

    """

    path: Path
    _run_in_executor: FunctionExecutor
    _run_subprocess: RunSubprocess
    upload: Callable[[Path, VirtoolFileFormat], Awaitable[None]] = not_implemented
    finalize: Callable[[], Awaitable[None]] = not_implemented
    _sequence_lengths: Optional[Dict[str, int]] = None
    _sequence_otu_map: Optional[Dict[str, str]] = None

    @property
    def bowtie_path(self) -> Path:
        """
        The path to the Bowtie2 index prefix for the Virtool index.

        """
        return self.path / "reference"

    @property
    def compressed_json_path(self) -> Path:
        """
        The path to the gzip-compressed JSON representation of the reference index in the workflow's work directory.

        """
        return self.path / "otus.json.gz"

    @property
    def fasta_path(self) -> Path:
        """
        The path to the complete FASTA file for the reference index in the workflow's work directory.

        """
        return self.path / "ref.fa"

    @property
    def json_path(self) -> Path:
        """
        The path to the JSON representation of the reference index in the workflow's work directory.

        """
        return self.path / "otus.json"

    async def decompress_json(self, processes: int):
        """
        Decompress the gzipped JSON file stored in the reference index directory. This data will be used to generate
        isolate indexes if required.

        Populate the ``sequence_otu_map`` attribute.

        :param processes: the number processes available for decompression

        """
        if self.json_path.is_file():
            raise FileExistsError("Index JSON file has already been decompressed")

        try:
            await self._run_in_executor(
                decompress_file, self.compressed_json_path, self.json_path
            )
        except gzip.BadGzipFile:
            await self._run_in_executor(
                shutil.copyfile, self.compressed_json_path, self.json_path
            )

        async with aiofiles.open(self.json_path) as f:
            data = json.loads(await f.read())

        sequence_lengths = dict()
        sequence_otu_map = dict()

        for otu in data:
            for isolate in otu["isolates"]:
                for sequence in isolate["sequences"]:
                    sequence_id = sequence["_id"]

                    sequence_otu_map[sequence_id] = otu["_id"]
                    sequence_lengths[sequence_id] = len(sequence["sequence"])

        self._sequence_lengths = sequence_lengths
        self._sequence_otu_map = sequence_otu_map

    def get_otu_id_by_sequence_id(self, sequence_id: str) -> str:
        """
        Return the OTU ID associated with the given ``sequence_id``.

        :param sequence_id: the sequence ID
        :return: the matching OTU ID

        """
        try:
            return self._sequence_otu_map[sequence_id]
        except KeyError:
            raise ValueError("The sequence_id does not exist in the index")

    def get_sequence_length(self, sequence_id: str) -> int:
        """
        Get the sequence length for the given ``sequence_id``.

        :param sequence_id: the sequence ID
        :return: the length of the sequence

        """
        try:
            return self._sequence_lengths[sequence_id]
        except KeyError:
            raise ValueError("The sequence_id does not exist in the index")

    async def write_isolate_fasta(
        self,
        otu_ids: List[str],
        path: Path,
        processes: int = 1,
    ) -> Dict[str, int]:
        """
        Generate a FASTA file for all of the isolates of the OTUs specified by ``otu_ids``.

        :param otu_ids: the list of OTU IDs for which to generate and index
        :param path: the path to the reference index directory
        :return: a dictionary of the lengths of all sequences keyed by their IDS

        """
        unique_otu_ids = set(otu_ids)

        async with aiofiles.open(self.json_path) as f:
            data = json.loads(await f.read())

        otus = [otu for otu in data if otu["_id"] in unique_otu_ids]

        lengths = dict()

        async with aiofiles.open(path, "w") as f:
            for otu in otus:
                for isolate in otu["isolates"]:
                    for sequence in isolate["sequences"]:
                        await f.write(f">{sequence['_id']}\n{sequence['sequence']}\n")
                        lengths[sequence["_id"]] = len(sequence["sequence"])

        await self._run_in_executor(
            compress_file, path, path.with_suffix(".fa.gz"), processes
        )

        return lengths

    async def build_isolate_index(
        self, otu_ids: List[str], path: Path, processes: int
    ) -> Tuple[Path, Dict[str, int]]:
        """
        Generate a FASTA file and Bowtie2 index for all of the isolates of the OTUs specified by ``otu_ids``.

        :param otu_ids: the list of OTU IDs for which to generate and index
        :param path: the path to the reference index directory
        :param processes: how many processes are available for external program calls
        :return: a tuple containing the path to the Bowtie2 index, FASTA files, and a dictionary of the lengths
            of all sequences keyed by their IDS

        """
        fasta_path = Path(f"{path}.fa")

        lengths = await self.write_isolate_fasta(otu_ids, fasta_path, processes)

        command = [
            "bowtie2-build",
            "--threads",
            str(processes),
            str(fasta_path),
            str(path),
        ]

        await self._run_subprocess(command)

        return fasta_path, lengths


@fixture
async def indexes(
    index_provider: IndexProvider,
    work_path: Path,
    proc: int,
    run_in_executor: FunctionExecutor,
    run_subprocess: RunSubprocess,
) -> List[Index]:
    """A workflow fixture that lists all reference indexes required for the workflow as :class:`.Index` objects."""
    index_ = await index_provider

    index_work_path = work_path / "indexes" / index_.id
    index_work_path.mkdir(parents=True, exist_ok=True)

    if index_.ready:
        await index_provider.download(index_work_path)
        index = Index(
            id=index_.id,
            manifest=index_.manifest,
            reference=index_.reference,
            path=index_work_path,
            ready=index_.ready,
            _run_in_executor=run_in_executor,
            _run_subprocess=run_subprocess,
        )
    else:
        await index_provider.download(index_work_path, "otus.json.gz")
        index = Index(
            id=index_.id,
            manifest=index_.manifest,
            reference=index_.reference,
            path=index_work_path,
            ready=index_.ready,
            upload=index_provider.upload,
            finalize=index_provider.finalize,
            _run_in_executor=run_in_executor,
            _run_subprocess=run_subprocess,
        )

    await index.decompress_json(proc)

    return [index]
