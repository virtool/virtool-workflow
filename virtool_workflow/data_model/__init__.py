from .analysis import Analysis
from .hmms import HMM
from .indexes import Index
from .jobs import Job, Status, State
from .references import Reference
from .samples import Sample
from .subtractions import Subtraction, NucleotideComposition

__all__ = [
    "Analysis",
    "HMM",
    "Index",
    "Job",
    "NucleotideComposition",
    "Reference",
    "Sample",
    "State",
    "Status",
    "Subtraction",
]
