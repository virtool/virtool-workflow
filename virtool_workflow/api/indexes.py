from pathlib import Path

import aiohttp
from virtool_core.models.index import Index

from virtool_workflow.api.errors import raising_errors_by_status_code
from virtool_workflow.api.utils import read_file_from_response, upload_file_via_put
from virtool_workflow.data_model.files import VirtoolFile, VirtoolFileFormat


class IndexProvider:
    """
    Provide access to Virtool's indexes via the Jobs API.

    :param index_id: The index ID for the current job.
    :param ref_id: The reference ID for the current job.
    :param index_path: The file system path to store index files.
    :param http: An :obj:`aiohttp.ClientSession` to use when making HTTP requests.
    :param jobs_api_connection_string: The base URL for the jobs API.
    """

    def __init__(
        self,
        index_id: str,
        ref_id: str,
        http: aiohttp.ClientSession,
        jobs_api_connection_string: str,
    ):
        self._index_id = index_id
        self._ref_id = ref_id
        self.http = http
        self.jobs_api_connection_string = jobs_api_connection_string

    async def get(self) -> Index:
        """Get the index for the current job."""
        async with self.http.get(
            f"{self.jobs_api_connection_string}/indexes/{self._index_id}"
        ) as response:
            async with raising_errors_by_status_code(response) as resp_json:
                return Index(**resp_json)

    async def upload(
        self, path: Path, format_: VirtoolFileFormat = "fasta"
    ) -> VirtoolFile:
        """
        Upload a file associated with the current Index.

        Allowed file names are:

            - reference.json.gz
            - reference.fa.gz
            - reference.1.bt2
            - reference.2.bt2
            - reference.3.bt2
            - reference.4.bt4
            - reference.rev.1.bt2
            - reference.rev.2.bt2

        :param path: The path to the file.
        :param format_: The format of the file.
        :return: A :class:`VirtoolFile` object.
        """
        return await upload_file_via_put(
            self.http,
            f"{self.jobs_api_connection_string}/indexes/{self._index_id}/files/{path.name}",
            path,
            format_,
        )

    async def download(self, target_path: Path, *names) -> Path:
        """Download files associated with the current index."""
        if not names:
            names = {
                "otus.json.gz",
                "reference.fa.gz",
                "reference.1.bt2",
                "reference.2.bt2",
                "reference.3.bt2",
                "reference.4.bt2",
                "reference.rev.1.bt2",
                "reference.rev.2.bt2",
            }

        for name in names:
            async with self.http.get(
                f"{self.jobs_api_connection_string}/indexes/{self._index_id}/files/{name}"
            ) as response:
                await read_file_from_response(response, target_path / name)

        return target_path

    async def finalize(self):
        """Finalize the current index."""
        async with self.http.patch(
            f"{self.jobs_api_connection_string}/indexes/{self._index_id}"
        ) as response:
            async with raising_errors_by_status_code(response) as resp_json:
                return Index(**resp_json)

    async def delete(self):
        """Delete an un-finished index."""
        async with self.http.delete(
            f"{self.jobs_api_connection_string}/indexes/{self._index_id}"
        ) as response:
            async with raising_errors_by_status_code(response, accept=[200, 204]):
                pass
