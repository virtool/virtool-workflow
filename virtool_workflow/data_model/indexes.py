from dataclasses import dataclass
from typing import Dict

from virtool_workflow.data_model.references import Reference


@dataclass
class Index:
    """A Virtool Index."""

    id: str
    manifest: Dict[str, int]
    reference: Reference
    ready: bool
