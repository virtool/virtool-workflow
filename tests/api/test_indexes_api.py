from pathlib import Path

import pytest

from tests.api.mocks.mock_index_routes import (
    TEST_INDEX_ID,
    TEST_INDEX,
    TEST_REF_ID,
)
from virtool_workflow.api.indexes import IndexProvider
from virtool_workflow.data_model import Reference
from virtool_workflow.data_model.files import VirtoolFile
from virtool_workflow.data_model.indexes import Index


@pytest.fixture
async def indexes_api(http, jobs_api_connection_string: str):
    return IndexProvider(TEST_INDEX_ID, TEST_REF_ID, http, jobs_api_connection_string)


async def test_get(indexes_api):
    index = await indexes_api

    assert isinstance(index, Index)
    assert "c93ec9a9" in index.manifest
    assert index.id == TEST_INDEX_ID
    assert isinstance(index.reference, Reference)
    assert index.reference.id == TEST_REF_ID


async def test_upload(indexes_api: IndexProvider, tmpdir):
    test_file = Path(tmpdir) / "reference.fa.gz"
    test_file.write_text("ACTGACG", encoding="utf-8")

    file = await indexes_api.upload(test_file)

    assert isinstance(file, VirtoolFile)
    assert file.name == "reference.fa.gz"
    assert file.format == "fasta"
    assert file.size == 7


async def test_download(indexes_api: IndexProvider, tmpdir):
    await indexes_api.download(Path(tmpdir))

    print(set(p.name for p in Path(tmpdir).iterdir()))

    assert set(p.name for p in Path(tmpdir).iterdir()) == {
        "otus.json.gz",
        "reference.fa.gz",
        "reference.1.bt2",
        "reference.2.bt2",
        "reference.3.bt2",
        "reference.4.bt2",
        "reference.rev.1.bt2",
        "reference.rev.2.bt2",
    }


async def test_finalize(indexes_api: IndexProvider):
    await indexes_api.finalize()

    assert TEST_INDEX["ready"] is True
