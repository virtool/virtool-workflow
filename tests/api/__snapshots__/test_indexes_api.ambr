# name: test_finalize
  dict({
    'change_count': 1419,
    'contributors': list([
      dict({
        'administrator': False,
        'count': 478,
        'handle': 'jasper',
        'id': '1kg24j7t',
      }),
      dict({
        'administrator': True,
        'count': 162,
        'handle': 'mrott',
        'id': 'ihvze2u9',
      }),
    ]),
    'created_at': datetime.datetime(2018, 2, 1, 0, 28, 49, 798000, tzinfo=datetime.timezone.utc),
    'files': list([
    ]),
    'has_files': False,
    'id': 'jiwncaqr',
    'job': dict({
      'id': 'wwssuhhy',
    }),
    'manifest': dict({
      'c93ec9a9': 0,
    }),
    'modified_otu_count': 4,
    'otus': list([
      dict({
        'change_count': 4,
        'id': 'd1efjvxk',
        'name': 'Agaricus bisporus virus 1',
      }),
      dict({
        'change_count': 6,
        'id': 'u9vbhtg1',
        'name': 'Alternaria solani chrysovirus 1',
      }),
      dict({
        'change_count': 5,
        'id': 'af3xznax',
        'name': 'Alternaria tenuissima partitivirus 1',
      }),
    ]),
    'ready': True,
    'reference': dict({
      'id': '21n3j5v6',
    }),
    'user': dict({
      'administrator': False,
      'handle': 'igboyes',
      'id': 'abc12345',
    }),
    'version': 0,
  })
# ---
# name: test_get
  dict({
    'change_count': 1419,
    'contributors': list([
      dict({
        'administrator': False,
        'count': 478,
        'handle': 'jasper',
        'id': '1kg24j7t',
      }),
      dict({
        'administrator': True,
        'count': 162,
        'handle': 'mrott',
        'id': 'ihvze2u9',
      }),
    ]),
    'created_at': datetime.datetime(2018, 2, 1, 0, 28, 49, 798000, tzinfo=datetime.timezone.utc),
    'files': list([
    ]),
    'has_files': False,
    'id': 'jiwncaqr',
    'job': dict({
      'id': 'wwssuhhy',
    }),
    'manifest': dict({
      'c93ec9a9': 0,
    }),
    'modified_otu_count': 4,
    'otus': list([
      dict({
        'change_count': 4,
        'id': 'd1efjvxk',
        'name': 'Agaricus bisporus virus 1',
      }),
      dict({
        'change_count': 6,
        'id': 'u9vbhtg1',
        'name': 'Alternaria solani chrysovirus 1',
      }),
      dict({
        'change_count': 5,
        'id': 'af3xznax',
        'name': 'Alternaria tenuissima partitivirus 1',
      }),
    ]),
    'ready': False,
    'reference': dict({
      'id': '21n3j5v6',
    }),
    'user': dict({
      'administrator': False,
      'handle': 'igboyes',
      'id': 'abc12345',
    }),
    'version': 0,
  })
# ---
